### ISSUE TITLE
EF Core 7 now warns about missing ValueComparer despite it being set 

### ISSUE THREAD

**Issue by DanielStout5** [dotnet/efcore#30139]
### Include your code

We are using this extension method to store columns as JSON:

```C#
        /// <summary>
        /// Marks a field as being stored as JSON in the database
        /// </summary>
        public static PropertyBuilder<T> IsJson<T>(this PropertyBuilder<T> bldr, JsonSerializerSettings settings = null)
        {
            var conv = new ValueConverter<T, string>(
                v => JsonConvert.SerializeObject(v, settings),
                v => JsonConvert.DeserializeObject<T>(v, settings)
            );

            var comp = new ValueComparer<T>(
                (l, r) => IsEqual(l, r),
                v => FindHashCode(v),
                v => JsonConvert.DeserializeObject<T>(JsonConvert.SerializeObject(v)));

            bldr.Metadata.SetValueComparer(comp);
            bldr.HasConversion(conv);

            return bldr;
        }
```

E.g.:
```C#
            builder.Entity<Order>(bld =>
            {
                bld.Property(x => x.TaxCharges)
                    .IsJson();
            });
```

After updating to EF Core 7, this warning is logged twice for every property using `IsJson` when the application is run

```
[12:28:41 WRN] Microsoft.EntityFrameworkCore.Model.Validation The property 'Order.TaxCharges' is a collection or enumeration type with a value converter but with no value comparer. Set a value comparer to ensure the collection/enumeration elements are compared correctly.
```

This is strange because we are setting the ValueComparer: `bldr.Metadata.SetValueComparer(comp);`

### Include provider and version information

EF Core version: 7.0.2
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: (e.g. .NET 6.0) .Net 7.0.2
Operating system: Windows
IDE: Visual Studio 2022 17.4.3

**Comment by DanielStout5** [comment]
Not entirely sure why, but it no longer logs the warning when I swap `SetValueComparer` and `HasConversion`, like so:

```
            bldr.HasConversion(conv);
            bldr.Metadata.SetValueComparer(comp);
```

And I noticed there's actually a method to set both at once, which does also work: `bldr.HasConversion(conv, comp);`

I'd like to know whether the change in EF Core 7 was that HasConversion has now started clearing the assigned ValueComparer, or if it always acted this way and is only just now warning about it.

I've poked through the EF Core code a bit and I don't actually see how HasConversion ends up clearing the ValueComparer..

**Comment by ajcvickers** [comment]
Note for triage: looks like the behavior has been like this since #22074, which was shipped in 5.0. But in 6.0 we were not generating the warning.

**Comment by ajcvickers** [comment]
Note from triage: we will leave the current behavior for HasConversion, and the warning is now correct.
